http-proxy-policy/policy/engine.py

```diff
@@ -0,0 +1,130 @@
+    """Apply rules to the request."""
+    result = _evaluate_rules(rules, request)
+    if not result:
+        request.status = models.PROXY_STATUS_PENDING
+        request.accepted_auth = None
+    elif result.verdict == models.Verdict.ACCEPT:
+        request.status = models.PROXY_STATUS_ACCEPTED
+        request.accepted_auth = result.auth
+    else:
+        request.status = models.PROXY_STATUS_REJECTED
+        request.accepted_auth = None
+    return request
+
+
+def _evaluate_rules(rules: list[models.Rule], request: models.Request) -> _EvaluationResult | None:
+    rules = [r for r in rules if r.requirer is None or r.requirer == request.requirer]
+    if not rules:
+        return None
+    deny_rules = [r for r in rules if r.verdict == models.Verdict.REJECT]
+    accept_rules = [r for r in rules if r.verdict == models.Verdict.ACCEPT]
+    denied_count = 0
+    for auth in request.auth:
+        if deny_rules and any(
+            _match_deny_rules(
+                deny_rules,
+                domain=domain,
+                auth=auth,
+                src_ips=request.src_ips,
+            )
+            for domain in request.domains
+        ):
+            denied_count += 1
+            continue
+        if accept_rules and all(
+            _match_accept_rules(
+                accept_rules,
+                domain=domain,
+                auth=auth,
+                src_ips=request.src_ips,
+            )
+            for domain in request.domains
+        ):
+            return _EvaluationResult(verdict=models.Verdict.ACCEPT, auth=auth)
+    if denied_count == len(request.auth):
+        return _EvaluationResult(verdict=models.Verdict.REJECT, auth=None)
+    return None
+
+
+def _match_deny_rules(
+    rules: list[models.Rule], domain: str, auth: str, src_ips: models.IpSet
+) -> bool:
+    if not rules:
+        return False
+    for rule in rules:
+        if rule.auth and auth not in rule.auth:
+            continue
+        if rule.domains and not any(
+            _domain_match(domain, rule_domain) for rule_domain in rule.domains
+        ):
+            continue
+        if rule.src_ips.is_empty() or rule.src_ips.overlap(src_ips):
+            return True
+    return False
+
+
+def _match_accept_rules(
+    rules: list[models.Rule], domain: str, auth: str, src_ips: models.IpSet
+) -> bool:
+    if not rules:
+        return False
+    rule_src_ips = models.IpSet([])
+    for rule in rules:
+        if rule.auth and auth not in rule.auth:
+            continue
+        if rule.domains and not any(
+            _domain_match(domain, rule_domain) for rule_domain in rule.domains
+        ):
+            continue
+        if rule.src_ips.is_empty():
+            return True
+        rule_src_ips += rule.src_ips
+    return rule_src_ips.is_superset_of(src_ips)
+
+
+def _domain_match(domain: str, rule_domain: str) -> bool:
+    return domain == rule_domain or rule_domain.endswith("." + domain)
```

Comment:
`rule_domain` and `domain` must be flipped in the endswith

Reply:
Updated, and I have added some more unit tests to verify this behavior.